apis/defaults.go

```diff
@@ -4,80 +4,88 @@ import (
+	log "github.com/platform9/nodeadm/pkg/logrus"
+	netutil "github.com/platform9/nodeadm/pkg/util/net"
+
+	"github.com/Jeffail/gabs"
 )
 
 // SetInitDefaults sets defaults on the configuration used by init
 func SetInitDefaults(config *InitConfiguration) {
-	// First set Networking defaults
-	SetNetworkingDefaults(&config.Networking)
-	// Second set MasterConfiguration.Networking defaults
-	SetMasterConfigurationNetworkingDefaultsWithNetworking(config)
-	// Third use the remainder of MasterConfiguration defaults
-	kubeadmv1alpha1.SetDefaults_MasterConfiguration(&config.MasterConfiguration)
-	config.MasterConfiguration.Kind = "MasterConfiguration"
-	config.MasterConfiguration.APIVersion = "kubeadm.k8s.io/v1alpha1"
-	config.MasterConfiguration.KubernetesVersion = constants.KubernetesVersion
-	config.MasterConfiguration.NoTaintMaster = true
-	addOrAppend(&config.MasterConfiguration.APIServerExtraArgs, "feature-gates", constants.FeatureGates)
-	addOrAppend(&config.MasterConfiguration.ControllerManagerExtraArgs, "feature-gates", constants.FeatureGates)
-	addOrAppend(&config.MasterConfiguration.SchedulerExtraArgs, "feature-gates", constants.FeatureGates)
 }
 
 // SetInitDynamicDefaults sets defaults derived at runtime
 func SetInitDynamicDefaults(config *InitConfiguration) error {
-	nodeName, err := constants.GetHostnameOverride()
-	if err != nil {
-		return fmt.Errorf("unable to dervice hostname override: %v", err)
+	if err := setAPIBindPort(config); err != nil {
+		return fmt.Errorf("unable to set a default value for masterConfiguration.api.bindPort: %s", err)
+	}
+	if err := setAPIControlPlaneEndpoint(config); err != nil {
+		return fmt.Errorf("unable to set a default value for masterConfiguration.api.controlPlaneEndpoint: %s", err)
+	}
+	if err := setKeepalivedInterface(config); err != nil {
+		return fmt.Errorf("unable to set a default value for vipConfiguraton.networkInterface: %s", err)
 	}
-	config.MasterConfiguration.NodeName = nodeName
 	return nil
 }
 
-// SetJoinDefaults sets defaults on the configuration used by join
-func SetJoinDefaults(config *JoinConfiguration) {
-	SetNetworkingDefaults(&config.Networking)
-}
-
-// SetNetworkingDefaults sets defaults for the network configuration
-func SetNetworkingDefaults(netConfig *Networking) {
-	if netConfig.ServiceSubnet == "" {
-		netConfig.ServiceSubnet = constants.DefaultServiceSubnet
+// setAPIBindPort sets the API BindPort, if it us not defined, to a default
+// value, because the port must be known to define the keepalived health check
+// script. If the VIP is not configured, setAPIBindPort does nothing.
+func setAPIBindPort(config *InitConfiguration) error {
+	if config.VIPConfiguration == nil {
+		return nil
+	}
+	p, err := gabs.Consume(config.MasterConfiguration)
+	if err != nil {
+		return fmt.Errorf("unable to parse kubeadm MasterConfiguration: %s", err)
 	}
-	if netConfig.DNSDomain == "" {
-		netConfig.DNSDomain = constants.DefaultDNSDomain
+	if p.ExistsP("api.bindPort") {
```

Comment:
Just adding a comment as per our discussion during code walk through. There is a tradeoff between knowing the type or using master configuration as just a blob of `map[string]{}`. Not knowing the type reemphasizes (and rightly so) the fact that nodeadm should just act as pass through of kubeadm's master configuration. However there might be cases like these where nodeadm would still have to be aware of partial if not entire schema of upstream type.

Reply:
This is the only case, and it's due to nodeadm configuring keepalived. I'd like to remove even this case in the future, if possible.